File: 0146-lru-cache/0146-lru-cache.java
class LRUCache {
    class CacheUnit {
        CacheUnit prev, next;
        int key, value;

        CacheUnit(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    CacheUnit head, tail;
    HashMap<Integer, CacheUnit> map;
    int capacity;
    int size;

    public LRUCache(int capacity) {
        head = new CacheUnit(0, 0);
        tail = new CacheUnit(0, 0);
        head.next = tail;
        tail.prev = head;

        map = new HashMap();

        this.capacity = capacity;
        size = 0;
    }
    
    void detach(CacheUnit cacheUnit) {
        cacheUnit.prev.next = cacheUnit.next;
        cacheUnit.next.prev = cacheUnit.prev;
    }

    void prepend(CacheUnit cacheUnit) {
        head.next.prev = cacheUnit;
        cacheUnit.next = head.next;
        head.next = cacheUnit;
        cacheUnit.prev = head;
    }

    public int get(int key) {
        if (!map.containsKey(key)) {
            return -1;
        }

        CacheUnit cacheUnit = map.get(key);
        if (head.next != cacheUnit) {
            detach(cacheUnit);
            prepend(cacheUnit);
        }

        return cacheUnit.value;
    }
    
    public void put(int key, int value) {
        CacheUnit cacheUnit;
        if (map.containsKey(key)) {
            cacheUnit = map.get(key);
            cacheUnit.value = value;
            
            if (head.next != cacheUnit) {
                detach(cacheUnit);
                prepend(cacheUnit);
            }
        } else {
            if (capacity > size) {
                cacheUnit = new CacheUnit(key, value);
                prepend(cacheUnit);
                size++;
            } else {
                cacheUnit = tail.prev;
                map.remove(cacheUnit.key);
                cacheUnit.key = key;
                cacheUnit.value = value;

                if (head.next != cacheUnit) {
                    detach(cacheUnit);
                    prepend(cacheUnit);
                }
            }
            
            map.put(key, cacheUnit);
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.py
class LRUCache:
    class Node:
        pass

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.allocated = 0
        self.head = None
        self.map = {}

    def printNodes(self) -> None:
        if None == self.head:
            return

        print("Linked nodes")
        node = self.head
        while True:
            print(node.key, node.value)
            node = node.next
            if self.head == node:
                break

    def move2end(self, node: Node) -> None:
        if self.head == node:
            self.head = self.head.next
            return

        node.next.prev = node.prev
        node.prev.next = node.next

        node.next = self.head
        node.prev = node.next.prev
        node.next.prev = node
        node.prev.next = node

    def get(self, key: int) -> int:
        node = self.map.get(key)
        if node == None:
            return -1

        self.move2end(node)
        #self.printNodes()
        return node.value

    def put(self, key: int, value: int) -> None:
        node = self.map.get(key)
        if node == None:
            if self.capacity > self.allocated:
                self.allocated += 1
                node = self.Node()
                if None == self.head:
                    self.head = node.next = node.prev = node
                else:
                    node.next = self.head
                    node.prev = node.next.prev
                    node.next.prev = node
                    node.prev.next = node
            else:
                node = self.head
                self.map.pop(node.key)
                self.head = self.head.next

            self.map[key] = node
            node.key = key
        else:
            self.move2end(node)

        node.value = value
        #self.printNodes()
        
# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)